package Reveal.Core.Messages.Data;

// The state of the system
// Note: may need additional configurations or subdivisions to map properly
message State {
  repeated double q = 1;
  repeated double dq = 2;
}

// A control to be applied
// Note: may need additional configurations or subdivisions to map properly
message Control {
  repeated double u = 1;
}

// The set of all scenarios.
message Digest {
  repeated Scenario scenario = 1;
}

message Experiment {
  optional string experiment_id = 1;
  optional string scenario_id = 2;
  optional int32 trials = 3;
}

// A given scenario to evaluate
message Scenario {
  required string id = 1;
  optional string description = 2 [default = ""];
  optional int32 trials = 3 [default = 0];
  repeated string uri = 4;
}

// A single trial within a scenario
message Trial {
  required string scenario_id = 1;
  required uint32 trial_id = 2;
  optional double t = 3;
  optional double dt = 4;
  optional State state = 5;
  optional Control control = 6;
}

// A single solution for a given trial within a scenario
message Solution {
  required string scenario_id = 1;
  required uint32 trial_id = 2;
  optional double t = 3;
  optional State state = 5;
}

// Robot protocol??

